### PR TITLE
citra-qt: Make hotkeys configurable via the GUI (Attempt 2)

### PR DESCRIPTION
**Pull request by adityaruplaha**
This is take 2 at this. Take 1 is at #3786. All its commits have been merged into the first one to make rebase easier. 
Tested on both MSVC & MinGW locally. All works & looks ok.

Original description is as follows.

---
Big thanks to @Kloen for letting me ~~steal~~ see & edit his code from Kloen/citra:hotkeys-once-again (originally #2773)!
**EDIT**: I've rewritten a lot of his code and have gone through the rest, so I can answer review questions regarding nearly all parts of it.

Pretty self explanatory. See the code for more details.
**Note: This doesn't allow hotkeys to be bound to controllers.**

---
# Major changes:
* Hotkeys separated into its own tab.
* Registration code for hotkeys has been rewritten by @Kloen.
* Config code for hotkeys has been almost completely rewritten.

# Current Issues:
All fixed!! :tada:
# Screenshots:
capture
For more, see the original PR.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4437)

### CODE REVIEW

**Comment by adityaruplaha** [comment]
Should I add the functionality of the missing hotkeys in this PR itself?

**Comment by zhaowenlan1779** [comment]
pls rebase

**Comment by adityaruplaha** [comment]
All right. Was planning to do it anyways, as I forgot to sign a commit.

**Comment by zhaowenlan1779** [comment]
Please resolve conflicts and I think we can try tagging this again :p

File: src/citra_qt/configuration/configure_dialog.cpp [hunk 1]
```diff
@@ -104,9 +115,11 @@ void ConfigureDialog::UpdateVisibleTabs() {
...
 
     const QHash<QString, QWidget*> widgets = {
-        {"General", ui->generalTab},   {"System", ui->systemTab}, {"Input", ui->inputTab},
-        {"Graphics", ui->graphicsTab}, {"Audio", ui->audioTab},   {"Camera", ui->cameraTab},
-        {"Debug", ui->debugTab},       {"Web", ui->webTab},       {"UI", ui->uiTab}};
+        {tr("General"), ui->generalTab},   {tr("System"), ui->systemTab},
+        {tr("Input"), ui->inputTab},       {tr("Hotkeys"), ui->hotkeysTab},
+        {tr("Graphics"), ui->graphicsTab}, {tr("Audio"), ui->audioTab},
+        {tr("Camera"), ui->cameraTab},     {tr("Debug"), ui->debugTab},
+        {tr("Web"), ui->webTab},           {tr("UI"), ui->uiTab}};
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_dialog.cpp]: Please do not translate them here

File: src/citra_qt/configuration/configure_hotkeys.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+void ConfigureHotkeys::applyConfiguration(HotkeyRegistry& registry) {
+    for (int key_id = 0; key_id < model->rowCount(); key_id++) {
+        QStandardItem* parent = model->item(key_id, 0);
+        for (int key_column_id = 0; key_column_id < parent->rowCount(); key_column_id++) {
+            QStandardItem* action = parent->child(key_column_id, 0);
+            QStandardItem* keyseq = parent->child(key_column_id, 1);
+            for (auto key_iterator = registry.hotkey_groups.begin();
+                 key_iterator != registry.hotkey_groups.end(); ++key_iterator) {
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_hotkeys.cpp]: I believe you can use `for (auto& [group, actions] : registry.hotkey_groups)` to make it more beautiful.
ditto below.

Comment (adityaruplaha) [reply src/citra_qt/configuration/configure_hotkeys.cpp]: I'm not really familiar with structured bindings, but I'll try to do so.

**Comment by Narugakuruga** [comment]
Canary 1121 b8f2371 Windows 10
After changing graphics settings while game runs, hotkeys become unavailable. For example, F9 and F10 can't change screen layout, Ctrl+Z can't unlock speed limit.

**Comment by adityaruplaha** [comment]
I'm investigating this issue. I haven't touched this part of the code, so
no idea why this is happening.

File: src/citra_qt/configuration/configure_hotkeys.cpp [hunk 3]
```diff
@@ -107,15 +107,13 @@ void ConfigureHotkeys::applyConfiguration(HotkeyRegistry& registry) {
...
-                            it2->second.keyseq = QKeySequence(keyseq->text());
-                        }
-                    }
+            for (auto& [group, sub_actions] : registry.hotkey_groups) {
+                if (group != parent->text())
+                    continue;
+                for (auto& [action_name, hotkey] : sub_actions) {
+                    if (action_name != action->text())
+                        continue;
+                    hotkey.keyseq = QKeySequence(keyseq->text());
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_hotkeys.cpp]: This can likely be a direct if

File: src/citra_qt/hotkeys.cpp [hunk 4]
```diff
@@ -12,47 +11,30 @@
...
+void HotkeyRegistry::LoadHotkeys() {
+    // Make sure NOT to use a reference here because it would become invalid once we call
+    // beginGroup()
+    for (auto shortcut : UISettings::values.shortcuts) {
+        Hotkey& hk = hotkey_groups[shortcut.group][shortcut.name];
+        if (!shortcut.shortcut.first.isEmpty()) {
+            hk.keyseq = QKeySequence::fromString(shortcut.shortcut.first, QKeySequence::NativeText);
+            hk.context = (Qt::ShortcutContext)shortcut.shortcut.second;
+        }
+        if (hk.shortcut)
```
[the remark this answers is not in the document]
Comment (zhaowenlan1779) [reply src/citra_qt/hotkeys.cpp]: I'm not convinced that's the proper fix. Why should we disconnect the shortcut on loading hotkeys? That should more likely belong to saving I think.

File: src/citra_qt/configuration/configure_hotkeys.cpp (hunk 3 above)
Comment (adityaruplaha) [reply src/citra_qt/configuration/configure_hotkeys.cpp]: I didn't want the vide to be over-indented. The earlier version was a bit hard to read because of that.

File: src/citra_qt/hotkeys.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (adityaruplaha) [reply src/citra_qt/hotkeys.cpp]: I'll investigate this when I can. But I may not get time for a while, so feel free to force push over to my branch if required.

**Comment by adityaruplaha** [comment]
Fixed. Requesting a review from @jroweboy & @B3n30.

File: src/citra_qt/hotkeys.cpp [hunk 5]
```diff
@@ -32,8 +32,10 @@ void HotkeyRegistry::LoadHotkeys() {
             hk.keyseq = QKeySequence::fromString(shortcut.shortcut.first, QKeySequence::NativeText);
             hk.context = (Qt::ShortcutContext)shortcut.shortcut.second;
         }
-        if (hk.shortcut)
+        if (hk.shortcut) {
+            hk.shortcut->disconnect();
```
Comment (zhaowenlan1779) [inline src/citra_qt/hotkeys.cpp]: Asking this again. Why is it needed to disconnect the shortcut when loading hotkeys? That does not make sense to me.

Comment (adityaruplaha) [reply src/citra_qt/hotkeys.cpp]: We should disconnect all the hotkeys because they will be connected again to their respective functions in main.cpp.

**Comment by zhaowenlan1779** [comment]
please rebase. thanks

File: src/citra_qt/configuration/configure_dialog.h [hunk 6]
```diff
@@ -17,12 +17,12 @@ class ConfigureDialog : public QDialog {
...
 
 public:
-    explicit ConfigureDialog(QWidget* parent, const HotkeyRegistry& registry);
+    explicit ConfigureDialog(QWidget* parent, HotkeyRegistry& registry);
     ~ConfigureDialog() override;
 
-    void applyConfiguration();
     void UpdateVisibleTabs();
     void PopulateSelectionList();
+    void applyConfiguration();
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_dialog.h]: is there a reason to move this?

File: src/citra_qt/configuration/configure_hotkeys.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+ConfigureHotkeys::ConfigureHotkeys(QWidget* parent)
+    : QWidget(parent), ui(std::make_unique<Ui::ConfigureHotkeys>()) {
+    ui->setupUi(this);
+    setFocusPolicy(Qt::ClickFocus);
+
+    model = new QStandardItemModel(this);
+    model->setColumnCount(3);
+    model->setHorizontalHeaderLabels({tr("Action"), tr("Hotkey"), tr("Context")});
+
+    ui->hotkey_list->setSelectionMode(QTreeView::SingleSelection);
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_hotkeys.cpp]: this, as well as the `setEditTriggers`, should be moved to the ui file if possible

File: src/citra_qt/configuration/configure_hotkeys.cpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    int return_code = hotkey_dialog->exec();
+
+    auto key_sequence = hotkey_dialog->GetSequence();
+
+    if (return_code == QDialog::Rejected || key_sequence.isEmpty())
+        return;
+
+    if (IsUsedKey(key_sequence) &&
+        key_sequence != QKeySequence(previous_key.toString(), QKeySequence::NativeText)) {
+        model->setData(index, previous_key);
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_hotkeys.cpp]: Is this necessary?

If I understandit correctly, the data is not changed yet at this point, right?

File: src/citra_qt/ui_settings.cpp [hunk 9]
```diff
@@ -14,5 +14,4 @@ const Themes themes{{
 }};
 
 Values values = {};
-
```
Comment (zhaowenlan1779) [inline src/citra_qt/ui_settings.cpp]: unrelated

**Comment by wwylele** [comment]
Temporarily untagged from canary to test #4538. Please resolve problems here soon so we can merge this into master

File: src/citra_qt/configuration/configure_hotkeys.cpp (hunk 7 above)
Comment (adityaruplaha) [reply src/citra_qt/configuration/configure_hotkeys.cpp]: I couldn't find a way to do this.

File: src/citra_qt/configuration/configure_hotkeys.cpp (hunk 8 above)
Comment (adityaruplaha) [reply src/citra_qt/configuration/configure_hotkeys.cpp]: Fixed locally. Will push along with more changes.

File: src/citra_qt/configuration/configure_hotkeys.cpp (hunk 7 above)
Comment (adityaruplaha) [reply src/citra_qt/configuration/configure_hotkeys.cpp]: Fixed locally.

File: src/citra_qt/configuration/configure_dialog.h (hunk 6 above)
Comment (adityaruplaha) [reply src/citra_qt/configuration/configure_dialog.h]: Fixed locally.

**Comment by adityaruplaha** [comment]
Fixed that issue. However, there is a minor nit:
>Shortcuts\Main%20Window\Exit%20Fullscreen\KeySeq\default=false
Shortcuts\Main%20Window\Exit%20Fullscreen\KeySeq=Esc

Couldn't find anything that would cause this. Changing the `/default` value to `true` and the KeySeq value to anything else restores it to Esc. Really shouldn't be much of an issue as most people won't try to hack the INI.
However, we won't be able to change defaults for that (& only that) hotkey.

**Comment by adityaruplaha** [comment]
Fixed that issue. Very big thanks to @wwylele for identifying the cause.

According to my tests, all works as expected.
- [x] Basic functionality
- [x] Config code
- [x] On-the-fly reloading

Please add `pr:needs-review`.

File: src/citra_qt/configuration/config.cpp [hunk 10]
```diff
@@ -48,6 +48,31 @@ const std::array<std::array<int, 5>, Settings::NativeAnalog::NumAnalogs> Config:
     },
 }};
 
+// This shouldn't have anything except static initializers (no functions). So
+// QKeySequnce(...).toString() is NOT ALLOWED HERE.
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/config.cpp]: This two sentences don't look too necessary to me...

File: src/citra_qt/configuration/configure_input.cpp [hunk 11]
```diff
@@ -230,6 +230,30 @@ void ConfigureInput::applyConfiguration() {
...
+    for (int button = 0; button < Settings::NativeButton::NumButtons; button++) {
+        auto button_param = buttons_param[button];
+
+        if (button_param.Get("engine", "") == "keyboard") {
+            list << QKeySequence(button_param.Get("code", 0));
+        }
+    }
+
+    // TODO(adityaruplaha): Add home button to list when we finally emulate it
+    // Button ID of home button is 14: Reffered from citra_qt/configuration/config.cpp
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_input.cpp]:
```suggestion
    // Button ID of home button is 14: Referred from citra_qt/configuration/config.cpp
```

File: src/citra_qt/hotkeys.h [hunk 12]
```diff
@@ -49,22 +48,26 @@ class HotkeyRegistry final {
     QShortcut* GetHotkey(const QString& group, const QString& action, QWidget* widget);
 
     /**
-     * Register a hotkey.
+     * Returns a QKeySequence object who signal can be connected to QAction->SetShortcut.
```
Comment (zhaowenlan1779) [inline src/citra_qt/hotkeys.h]:
```suggestion
     * Returns a QKeySequence object whose signal can be connected to QAction::setShortcut.
```

File: src/citra_qt/hotkeys.h [hunk 13]
```diff
@@ -49,22 +48,26 @@ class HotkeyRegistry final {
...
-     *                        file before
-     * @warning Both the group and action strings will be displayed in the hotkey settings dialog
+     * @param group  General group this hotkey belongs to (e.g. "Main Window", "Debugger").
+     * @param action Name of the action (e.g. "Start Emulation", "Load Image").
+     */
+    QKeySequence GetKeySequence(const QString& group, const QString& action);
+
+    /**
+     * Returns a Qt::ShortcutContext object who can be connected to other
+     * QAction->SetShortcutContext.
```
Comment (zhaowenlan1779) [inline src/citra_qt/hotkeys.h]:
```suggestion
     * QAction::setShortcutContext.
```

File: src/citra_qt/hotkeys.h [hunk 14]
```diff
@@ -49,22 +48,26 @@ class HotkeyRegistry final {
...
+     *
+     * @param group  General group this shortcutcontext belongs to (e.g. "Main Window", "Debugger").
```
Comment (zhaowenlan1779) [inline src/citra_qt/hotkeys.h]:
```suggestion
     * @param group  General group this shortcut context belongs to (e.g. "Main Window", "Debugger").
```

File: src/citra_qt/hotkeys.cpp [hunk 15]
```diff
@@ -12,47 +11,32 @@
...
-    if (hotkey_group.find(action) != hotkey_group.end()) {
-        return;
+void HotkeyRegistry::LoadHotkeys() {
+    // Make sure NOT to use a reference here because it would become invalid once we call
+    // beginGroup()
+    for (auto shortcut : UISettings::values.shortcuts) {
+        Hotkey& hk = hotkey_groups[shortcut.group][shortcut.name];
+        if (!shortcut.shortcut.first.isEmpty()) {
+            hk.keyseq = QKeySequence::fromString(shortcut.shortcut.first, QKeySequence::NativeText);
+            hk.context = (Qt::ShortcutContext)shortcut.shortcut.second;
```
Comment (zhaowenlan1779) [inline src/citra_qt/hotkeys.cpp]:
```suggestion
            hk.context = static_cast<Qt::ShortcutContext>(shortcut.shortcut.second);
```

**Comment by wwylele** [comment]
Sorry, please resolve conflict again 😅

File: src/citra_qt/configuration/config.cpp (hunk 10 above)
Comment (adityaruplaha) [reply src/citra_qt/configuration/config.cpp]: I thought I should've explained why can't it be used. It will cause a real bad problem if it's used. Especially if a debugger is unavailable.

File: src/citra_qt/util/sequence_dialog/sequence_dialog.h [hunk 16]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2018 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <memory>
+#include <QDialog>
+#include <QKeySequenceEdit>
```
Comment (zhaowenlan1779) [inline src/citra_qt/util/sequence_dialog/sequence_dialog.h]: This can likely be a forward declaration, and you can include `<QKeySequence>` instead.

File: src/citra_qt/configuration/configure_hotkeys.h [hunk 17]
```diff
@@ -0,0 +1,59 @@
+// Copyright 2017 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <memory>
+#include <QStandardItemModel>
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_hotkeys.h]: This can likely be a forward declaration

File: src/citra_qt/configuration/configure_hotkeys.h [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+#include <QWidget>
+#include "common/param_package.h"
```
Comment (zhaowenlan1779) [inline src/citra_qt/configuration/configure_hotkeys.h]: Where is this used?

**Comment by adityaruplaha** [comment]
Sorry I didn't get much time to address review comments but I've fixed the build.

File: src/citra_qt/util/sequence_dialog/sequence_dialog.h (hunk 16 above)
Comment (adityaruplaha) [reply src/citra_qt/util/sequence_dialog/sequence_dialog.h]: > you can include `<QKeySequence>` instead.

That's not required.

**Comment by adityaruplaha** [comment]
Fixed everything. Now is there any hope for it being merged?

**Comment by wwylele** [comment]
Will merge soon if no more comments
